File: leet_code/trees/bfs.rb
def bfs(root)
    return [] if !root
    result, queue = [], [root]
    until queue.empty?
        height, length = 1, queue.length
        subarr = []
        while height <= length
            node = queue.shift
            subarr << node.val
            queue << node.left if node.left
            queue << node.right if node.right
            height += 1
        end
        result << subarr
    end
    result
end

File: leet_code/trees/bfs.js
const bfs = root => {
    if (!root) return [];
    const result = [],
          queue = [root];
    while (queue.length > 0) {
        const subArr = [],
              length = queue.length;
        for (let i = 1; i <= length; i++) {
            const node = queue.shift();
            subArr.push(node.val);
            if (node.left) queue.push(node.left);
            if (node.right) queue.push(node.right);
        }
        result.push(subArr);
    }
    return result;
};
